File: libs/core/src/declarations/declaration_tree_builder.cpp
/**
 * @file declaration_tree_builder.cpp
 * @author Francesco Corso
 * @date 8/16/2026
 * @brief
 */
module;

#include <libassert/assert-macros.hpp>

module prism.core:declarations.declaration_tree_builder.impl;

import :declarations.declaration_tree_builder;
import :syntax.visit;
import :syntax.tree;
import :syntax.green.visit;
import :declarations.single_root_namespace_declaration;
import :diagnostics.diagnostic_bag;
import :memory.buffer_pool;
import :binder.binding_helpers;

namespace prism
{

    DeclarationTreeBuilder::DeclarationTreeBuilder(const SyntaxTree &syntax_tree) : syntax_tree_{syntax_tree}
    {
    }

    RefCountPtr<const SingleRootNamespaceDeclaration> DeclarationTreeBuilder::for_tree(const SyntaxTree &syntax_tree)
    {
        DeclarationTreeBuilder builder{std::move(syntax_tree)};
        return builder.visit_tree();
    }

    RefCountPtr<const SingleRootNamespaceDeclaration> DeclarationTreeBuilder::visit_tree()
    {
        return visit(syntax_tree_.root(),
                     Overload{[this](const CompilationUnitSyntax &syntax) { return visit_compilation_unit(syntax); },
                              [](const SyntaxNode &syntax) -> RefCountPtr<const SingleRootNamespaceDeclaration>
                              {
                                  throw InvalidStateException{
                                      std::format("Unexpected syntax node type: {}", get_name(syntax.kind()))};
                              }});
    }

    ImmutableArray<RefCountPtr<const SingleDeclaration>> DeclarationTreeBuilder::visit_namespace_children(
        const SyntaxNode &node,
        SyntaxList<DeclarationSyntax> members)
    {
        DEBUG_ASSERT(node.is<NamespaceDeclarationSyntax>() || node.is<CompilationUnitSyntax>());

        if (members.empty())
            return {};

        PooledVector<RefCountPtr<const SingleDeclaration>> builder;
        for (auto &member : members)
        {
            auto namespace_or_type =
                visit(member,
                      Overload{[this](const NamespaceDeclarationSyntax &syntax)
                               { return visit_namespace_declaration(syntax); },
                               [](const DeclarationSyntax &) -> RefCountPtr<const SingleDeclaration>
                               {
                                   return nullptr;
                               }});
            if (namespace_or_type != nullptr)
                builder.push_back(std::move(namespace_or_type));
        }

        return ImmutableArray{std::from_range, builder | std::views::as_rvalue};
    }

    RefCountPtr<const SingleRootNamespaceDeclaration> DeclarationTreeBuilder::visit_compilation_unit(
        const CompilationUnitSyntax &compilation_unit)
    {
        auto children = visit_namespace_children(compilation_unit, compilation_unit.members());
        return create_single_root_namespace_declaration(compilation_unit, std::move(children));
    }

    RefCountPtr<const SingleDeclaration> DeclarationTreeBuilder::visit_namespace_declaration(
        const NamespaceDeclarationSyntax &node)
    {
        auto children = visit_namespace_children(node, node.members());

        auto has_usings = !node.usings().empty();
        auto *name = &node.name();
        const SyntaxNode *current_node = &node;
        auto member_names = get_non_type_member_names(node.members());
        while (name->is<QualifiedNameSyntax>())
        {
            auto qualified = static_cast<const QualifiedNameSyntax *>(name);
            auto right_name = get_unqualified_name(qualified->right());
            auto ns = make_ref_counted<const SingleNamespaceDeclaration>(
                right_name,
                has_usings,
                SyntaxReference{*current_node},
                SourceLocation{qualified->right()},
                std::move(children),
                std::exchange(member_names, ImmutableHashSet<Name>{}),
                ImmutableArray<Diagnostic>{});

            children = ImmutableArray<RefCountPtr<const SingleDeclaration>>{std::move(ns)};
            current_node = &qualified->left();
            has_usings = false;
        }

        DiagnosticBag diagnostics{};

        if (node.is<FileScopedNamespaceDeclarationSyntax>())
        {
            if (is_instance<FileScopedNamespaceDeclarationSyntax>(node.parent()))
            {
                diagnostics.add(Diagnostic{DiagnosticInfo::create<DiagnosticCode::multiple_file_scoped_namespaces>(),
                                           node.name().location()});
            }
            else if (is_instance<NamespaceDeclarationSyntax>(node.parent()))
            {
                diagnostics.add(Diagnostic{DiagnosticInfo::create<DiagnosticCode::file_scoped_and_normal_namespace>(),
                                           node.name().location()});
            }
            else
            {
                DEBUG_ASSERT(is_instance<CompilationUnitSyntax>(node.parent()));
                auto &compilation_unit = static_cast<const CompilationUnitSyntax &>(*node.parent());
                if (&node != &compilation_unit.members()[0])
                {
                    diagnostics.add(Diagnostic{
                        DiagnosticInfo::create<DiagnosticCode::file_scoped_namespace_not_before_all_members>(),
                        node.name().location()});
                }
            }
        }
        else
        {
            DEBUG_ASSERT(node.is<BlockNamespaceDeclarationSyntax>());

            if (is_instance<FileScopedNamespaceDeclarationSyntax>(node.parent()))
            {
                diagnostics.add(Diagnostic{DiagnosticInfo::create<DiagnosticCode::file_scoped_and_normal_namespace>(),
                                           node.name().location()});
            }
        }

        if (!node.modifiers().empty())
        {
            diagnostics.add(Diagnostic{DiagnosticInfo::create<DiagnosticCode::bad_modifiers_on_namespace>(),
                                       node.modifiers()[0].location()});
        }

        DEBUG_ASSERT(name->is<SimpleNameSyntax>());
        auto unqualified_name = get_unqualified_name(static_cast<const SimpleNameSyntax &>(*name));
        return make_ref_counted<SingleNamespaceDeclaration>(unqualified_name,
                                                            has_usings,
                                                            SyntaxReference{*current_node},
                                                            SourceLocation{*name},
                                                            std::move(children),
                                                            std::move(member_names),
                                                            std::move(diagnostics).move_to<ImmutableArray>());
    }

    RefCountPtr<const SingleRootNamespaceDeclaration> DeclarationTreeBuilder::create_single_root_namespace_declaration(
        const CompilationUnitSyntax &compilation_unit,
        ImmutableArray<RefCountPtr<const SingleDeclaration>> children)
    {
        bool has_usings = !compilation_unit.usings().empty();

        return make_ref_counted<const SingleRootNamespaceDeclaration>(
            has_usings,
            SyntaxReference{compilation_unit},
            std::move(children),
            get_non_type_member_names(compilation_unit.members()),
            ImmutableArray<Diagnostic>{});
    }

    ImmutableHashSet<Name> DeclarationTreeBuilder::get_non_type_member_names(SyntaxList<DeclarationSyntax> members)
    {
        std::unordered_set<Name> member_names;
        for (auto &member : members)
        {
            visit(member,
                  Overload{[&](const VariableDeclarationSyntax &variable)
                           { member_names.insert(get_identifier_name(variable.identifier())); },
                           [&](const FunctionDeclarationSyntax &function)
                           { member_names.insert(get_identifier_name(function.identifier())); },
                           [&](const DeclarationSyntax &)
                           {
                               // Other declarations aren't part of this
                           }});
        }

        return ImmutableHashSet{std::from_range, member_names};
    }
} // namespace prism

File: src/Prism.Core/Declarations/DeclarationTreeBuilder.cs
using System.Collections.Immutable;
using System.Diagnostics;
using Prism.Core.Binding;
using Prism.Core.Diagnostics;
using Prism.Core.Syntax;

namespace Prism.Core.Declarations;

internal sealed class DeclarationTreeBuilder
{
    private readonly SyntaxTree _syntaxTree;
    private int _currentTypeIndex;

    private DeclarationTreeBuilder(SyntaxTree syntaxTree)
    {
        _syntaxTree = syntaxTree;
    }

    public static SingleRootNamespaceDeclaration ForTree(SyntaxTree tree)
    {
        var builder = new DeclarationTreeBuilder(tree);
        return builder.VisitTree();
    }

    private SingleRootNamespaceDeclaration VisitTree()
    {
        return _syntaxTree.Root is CompilationUnitSyntax compilationUnitSyntax
            ? VisitCompilationUnit(compilationUnitSyntax)
            : throw new InvalidOperationException("Compilation unit syntax expected");
    }

    private ImmutableArray<SingleDeclaration> VisitNamespaceChildren(
        SyntaxNode node,
        SyntaxList<DeclarationSyntax> members
    )
    {
        Debug.Assert(node is CompilationUnitSyntax or NamespaceDeclarationSyntax);

        if (members.Count == 0)
            return [];

        var builder = ImmutableArray.CreateBuilder<SingleDeclaration>();
        foreach (var member in members)
        {
            var namespaceOrType = member switch
            {
                NamespaceDeclarationSyntax ns => VisitNamespaceDeclaration(ns),
                _ => null,
            };

            if (namespaceOrType is not null)
                builder.Add(namespaceOrType);
        }

        return builder.DrainToImmutable();
    }

    private SingleRootNamespaceDeclaration VisitCompilationUnit(CompilationUnitSyntax syntax)
    {
        var children = VisitNamespaceChildren(syntax, syntax.Members);
        return CreateSingleRootNamespaceDeclaration(syntax, children);
    }

    private SingleDeclaration VisitNamespaceDeclaration(NamespaceDeclarationSyntax node)
    {
        var children = VisitNamespaceChildren(node, node.Members);

        var hasUsings = node.Usings.Count > 0;
        var name = node.Name;
        SyntaxNode? currentNode = node;
        var memberNames = GetNonTypeMemberNames(node.Members);
        while (name is QualifiedNameSyntax qualified)
        {
            var rightName = qualified.UnqualifiedName;
            var ns = new SingleNamedNamespaceDeclaration(
                rightName,
                new SyntaxReference(currentNode),
                new SourceLocation(qualified.Right),
                memberNames,
                children
            )
            {
                HasUsings = hasUsings,
            };

            memberNames = ImmutableHashSet<string>.Empty;
            children = [ns];
            name = qualified.Left;
            currentNode = name;
            hasUsings = false;
        }

        var diagnostics = DiagnosticBag.Create();
        switch (node)
        {
            case FileScopedNamespaceDeclarationSyntax:
                switch (node.Parent)
                {
                    case FileScopedNamespaceDeclarationSyntax:
                        diagnostics.Add(
                            Diagnostic.MultipleFileScopedNamespaces(node.Name.Location)
                        );
                        break;
                    case NamespaceDeclarationSyntax:
                        diagnostics.Add(
                            Diagnostic.FileScopedAndNormalNamespace(node.Name.Location)
                        );
                        break;
                    case CompilationUnitSyntax compilationUnit:
                    {
                        if (!ReferenceEquals(node, compilationUnit.Members[0]))
                        {
                            diagnostics.Add(
                                Diagnostic.FileScopedNamespaceNotBeforeAllMembers(
                                    node.Name.Location
                                )
                            );
                        }

                        break;
                    }
                    default:
                        throw new InvalidOperationException("This isn't a valid state");
                }

                break;
            case BlockNamespaceDeclarationSyntax:
            {
                if (node.Parent is FileScopedNamespaceDeclarationSyntax)
                {
                    diagnostics.Add(Diagnostic.FileScopedAndNormalNamespace(node.Name.Location));
                }

                break;
            }
            default:
                throw new InvalidOperationException("This isn't a valid state");
        }

        if (node.Modifiers.Count > 0)
        {
            diagnostics.Add(Diagnostic.BadModifiersOnNamespace(node.Modifiers[0].Location));
        }

        Debug.Assert(name is SimpleNameSyntax);
        var unqualifiedName = name.UnqualifiedName;
        return new SingleNamedNamespaceDeclaration(
            unqualifiedName,
            new SyntaxReference(currentNode),
            new SourceLocation(name),
            memberNames,
            children
        )
        {
            HasUsings = hasUsings,
            Diagnostics = diagnostics.ToImmutableAndClear(),
        };
    }

    private SingleRootNamespaceDeclaration CreateSingleRootNamespaceDeclaration(
        CompilationUnitSyntax syntax,
        ImmutableArray<SingleDeclaration> children
    )
    {
        return new SingleRootNamespaceDeclaration(
            new SyntaxReference(syntax),
            new SourceLocation(syntax),
            GetNonTypeMemberNames(syntax.Members),
            children
        )
        {
            HasUsings = syntax.Usings.Count > 0,
        };
    }

    private static ImmutableHashSet<string> GetNonTypeMemberNames(
        SyntaxList<DeclarationSyntax> members
    )
    {
        var memberNames = ImmutableHashSet.CreateBuilder<string>();
        foreach (var member in members)
        {
            switch (member)
            {
                case VariableDeclarationSyntax variable:
                    memberNames.Add(variable.Identifier.IdentifierName);
                    break;
                case FunctionDeclarationSyntax function:
                    memberNames.Add(function.Identifier.IdentifierName);
                    break;
            }
        }

        return memberNames.ToImmutable();
    }
}
